### PR TITLE
Have default values in add database dialog input boxes

### LINKED ISSUE #12041: Add validation for SQLCMD variables in database reference dialog
Need to check for invalid characters and add $() around the variable if necessary. Show error message in Example usage section if name is invalid.

### PR DESCRIPTION
**Pull request by kisantia**
This adds default values for the database and server name and variable input boxes:
defaultValues2

Also update the example usage to handle if the sqlcmd variables are in the $() or not and show the example usage accordingly. If the input box value doesn't have $() around the variable, the example usage shows it as $(variable), and if the variable is already in the format $(variable), it doesn't get added. Also added validation for invalid sqlcmd variable names(fixes #12041)
image
image

### CODE REVIEW

File: extensions/sql-database-projects/src/dialogs/addDatabaseReferenceDialog.ts [hunk 1]
```diff
@@ -327,11 +335,38 @@ export class AddDatabaseReferenceDialog {
...
+			this.setDefaultDatabaseValues();
+			this.serverNameTextbox!.value = constants.otherServer;
+			this.serverVariableTextbox!.value = constants.otherSeverVariable;
+		}
+	}
+
+	private setDefaultDatabaseValues(): void {
+		switch (this.currentReferenceType) {
+			case ReferenceType.project: {
+				// TODO: add when projects support is added
```
Comment (Benjin) [inline extensions/sql-database-projects/src/dialogs/addDatabaseReferenceDialog.ts:357]: Can you make this into a GH issue and add the ID here?

File: extensions/sql-database-projects/src/common/utils.ts [hunk 2]
```diff
@@ -143,3 +143,74 @@ export function readSqlCmdVariables(xmlDoc: any): Record<string, string> {
...
+	return name.trim();
+}
+
+/**
+ * 	Format as sqlcmd variable by adding $() if necessary
+ * if the variable already starts with $(, then add )
+ * @param name
+ */
+export function formatSqlCmdVariable(name: string): string {
+	// TODO: check if it's a valid variable name and just return if it isn't
```
Comment (Benjin) [inline extensions/sql-database-projects/src/common/utils.ts:176]: Maybe throw instead of returning empty string if it's not?

File: extensions/sql-database-projects/src/common/utils.ts [hunk 3]
```diff
@@ -143,3 +143,74 @@ export function readSqlCmdVariables(xmlDoc: any): Record<string, string> {
...
+
+	return name;
+}
+
+/**
+ * Checks if it's a valid sqlcmd variable name
+ * https://docs.microsoft.com/en-us/sql/ssms/scripting/sqlcmd-use-with-scripting-variables?redirectedfrom=MSDN&view=sql-server-ver15#guidelines-for-scripting-variable-names-and-values
+ * @param name variable name to validate
+ */
+export function isValidSqlCmdVariableName(name: string | undefined): boolean {
```
Comment (Benjin) [inline extensions/sql-database-projects/src/common/utils.ts:199]: If this throws instead of returning a boolean, you can give an explanation to the user as to why their variable name isn't valid.

File: extensions/sql-database-projects/src/dialogs/addDatabaseReferenceDialog.ts (hunk 1 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/dialogs/addDatabaseReferenceDialog.ts:357]: I'll add this in this PR after #12186 is merged

File: extensions/sql-database-projects/src/common/constants.ts [hunk 4]
```diff
@@ -115,6 +115,8 @@ export const exampleUsage = localize('exampleUsage', "Example Usage");
 export const enterSystemDbName = localize('enterSystemDbName', "Enter a database name for this system database");
 export const databaseNameRequiredVariableOptional = localize('databaseNameRequiredVariableOptional', "A database name is required. The database variable is optional.");
 export const databaseNameServerNameVariableRequired = localize('databaseNameServerNameVariableRequired', "A database name, server name, and server variable are required. The database variable is optional");
+export const otherServer = localize('otherServer', "OtherServer");
+export const otherSeverVariable = localize('otherServerVariable', "$(OtherServer)");
```
[the remark this answers is not in the document]
Comment (kisantia) [reply extensions/sql-database-projects/src/common/constants.ts:119]: SSDT allows it, but then there's an error when publishing if you have a variable name like $(ðĶƢ). Should we check for this or just let publishing fail?

File: extensions/sql-database-projects/src/common/utils.ts (hunk 2 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/common/utils.ts]: I removed this comment since I added it before I figured out how the validation stuff would work.

File: extensions/sql-database-projects/src/common/utils.ts (hunk 3 above)
Comment (kisantia) [reply extensions/sql-database-projects/src/common/utils.ts:198]: created https://github.com/microsoft/azuredatastudio/issues/12231. For now, this is just the same as what SSDT does for invalid variable names

**Review by udeeshagautam — APPROVED** [review]
(re signing off my connection got interrupted)

Looks good. I might be missing some edge cases, will test with your change tomorrow.. but don't be blocked on me..

File: extensions/sql-database-projects/src/common/constants.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Benjin) [reply extensions/sql-database-projects/src/common/constants.ts:123]: I think just let publishing fail; if someone hits it and cares, they can submit a dacfx bug.  This way, we don't have to come back and change logic in db projects if that happens and we fix it.

The main reason I asked was that maybe localizing "OtherServer" would signal to users that we encourage them to use localized variable names which, since they're unicode, would fail during publish.  I think maybe we shouldn't localize these two.

**Review by Benjin — APPROVED** [review]

File: extensions/sql-database-projects/src/common/constants.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (kisantia) [reply extensions/sql-database-projects/src/common/constants.ts:123]: makes sense. I'll change them to not be localized

**Review by SakshiS-harma — APPROVED** [review]

File: extensions/sql-database-projects/src/test/dialogs/addDatabaseReferenceDialog.test.ts [hunk 5]
```diff
@@ -32,35 +32,36 @@ describe('Add Database Reference Dialog', () => {
...
+		should(dialog.databaseNameTextbox!.value).equal('', 'database name text box should be empty because no dacpac has been selected');
 		should(dialog.dialog.okButton.enabled).equal(false, 'Ok button should not be enabled because dacpac input box is not filled');
 
-		// fill in dacpac textbox
+		// fill in dacpac textbox and database name text box
 		dialog.dacpacTextbox!.value = 'testDb.dacpac';
+		dialog.databaseNameTextbox!.value = 'testDb';
 		dialog.tryEnableAddReferenceButton();
-		should(dialog.dialog.okButton.enabled).equal(true, 'Ok button should be enabled after the dacpac textbox is filled');
+		should(dialog.dialog.okButton.enabled).equal(true, `Ok button should be enabled after the dacpac textbox is filled`);
```
[the remark this answers is not in the document]
Comment (kisantia) [reply extensions/sql-database-projects/src/test/dialogs/addDatabaseReferenceDialog.test.ts:58]: oops that's from when I was trying to figure out why the test was failing
